cou/steps/analyze.py

```diff
@@ -0,0 +1,256 @@
+            os_sequence = sorted(
+                list(self.os_version_units),
+                key=CompareOpenStack,
+            )
+            for os_release in os_sequence[:-1]:
+                next_release = determine_next_openstack_release(os_release)[1]
+                upgrade_units[next_release].update(self.os_release_units[os_release])
+                logging.warning(
+                    "upgrade units: %s from: %s to %s",
+                    self.os_release_units[os_release],
+                    os_release,
+                    next_release,
+                )
+        return upgrade_units
+
+    def check_os_channels_and_migration(
+        self, change_channel: defaultdict(set), charmhub_migration: defaultdict(set)
+    ) -> Tuple:
+        if len(self.os_version_units) > 1:
+            logging.warning(
+                "Skip check of channels. App %s has units with different openstack version",
+                self.name,
+            )
+        else:
+            actual_release = list(self.os_version_units)[0]
+            expected_channel = f"{actual_release}/stable"
+            if actual_release not in self.channel:
+                change_channel[expected_channel].add(self.name)
+                logging.warning(
+                    "App:%s need to track the channel: %s", self.name, expected_channel
+                )
+            if self.origin == "cs":
+                charmhub_migration[expected_channel].add(self.name)
+                logging.warning("App:%s need to perform migration to charmhub", self.name)
+        return change_channel, charmhub_migration
+
+    def check_os_origin(self, change_openstack_release: defaultdict(set)) -> defaultdict(set):
+        os_charm_config = self.get_os_origin()
+        if len(self.os_version_units) > 1:
+            logging.warning(
+                "Skip openstack-origin check. App %s has units with different openstack version",
+                self.name,
+            )
+        else:
+            actual_release = list(self.os_version_units)[0]
+            expected_os_origin = f"cloud:focal-{actual_release}"
+            # Exceptionally, if upgrading from Ussuri to Victoria
+            if actual_release == "ussuri":
+                if os_charm_config != "distro":
+                    logging.warning(
+                        "App: %s need to set openstack-origin or source to 'distro'", self.name
+                    )
+                    change_openstack_release["distro"].add(self.name)
+            else:
+                if expected_os_origin not in os_charm_config:
+                    change_openstack_release[expected_os_origin].add(self.name)
+                    logging.warning(
+                        "App: %s need to set openstack-origin or source to %s",
+                        self.name,
+                        expected_os_origin,
+                    )
+        return change_openstack_release
+
+
+def get_openstack_release(unit, model_name):
```

Comment:
the more I think of this, the more centralizing version checks seems wrong to me: charms are inherently decentralized, why are we ssh-ing into the units? determining the version of an app should be determined by the app itself; the upgrader should simply ask the charm

Reply:
I left a comment in the code saying that Ideally, the application should provide the openstack version and packages versions by a charm action and this might be possible with Sumbeam.